File: src/LabLink.Domain/Authorization/Permissions.cs
namespace LabLink.Domain.Authorization;

/// <summary>Danh mục permission key — nguồn sự thật ở server.
/// Phải khớp với frontend (auth/permissions.ts).</summary>
public static class Permissions
{
    public const string CatalogRead = "catalog.read";
    public const string CatalogPriceEdit = "catalog.price.edit";
    public const string DealCreate = "deal.create";
    public const string DealRead = "deal.read";
    public const string DealApprove = "deal.approve";
    public const string OrderCreate = "order.create";
    public const string OrderRead = "order.read";
    public const string SampleCollect = "sample.collect";   // B2 · Lấy mẫu
    public const string SampleSend = "sample.send";
    public const string SampleGather = "sample.gather";     // B3 · Gom mẫu
    public const string SampleReceive = "sample.receive";   // B4 · Nhận mẫu
    public const string SampleQc = "sample.qc";
    public const string HardcopyDeliver = "hardcopy.deliver";   // B6 · Giao bản cứng
    public const string HardcopyReceive = "hardcopy.receive";   // B7 · Nhận bản cứng
    public const string SidPrint = "sid.print";
    public const string ResultRead = "result.read";
    public const string ResultUpload = "result.upload";
    public const string ResultVerify = "result.verify";
    public const string UserManage = "user.manage";
    public const string RoleManage = "role.manage";
    public const string AuditRead = "audit.read";
    public const string ReportRead = "report.read";

    public static readonly string[] All =
    {
        CatalogRead, CatalogPriceEdit,
        DealCreate, DealRead, DealApprove,
        OrderCreate, OrderRead,
        SampleCollect, SampleSend, SampleGather, SampleReceive, SampleQc,
        HardcopyDeliver, HardcopyReceive,
        SidPrint,
        ResultRead, ResultUpload, ResultVerify,
        UserManage, RoleManage, AuditRead, ReportRead,
    };
}

/// <summary>Nhóm mặc định + tập quyền (README · RBAC · Role mặc định).</summary>
public static class DefaultRoles
{
    public record RoleSeed(string Code, string Name, string[] Permissions);

    public static readonly RoleSeed[] All =
    {
        new("doctor", "Bác sĩ chỉ định", new[]
        {
            Permissions.CatalogRead,
            Permissions.DealCreate, Permissions.DealRead,
            Permissions.OrderCreate, Permissions.OrderRead,
            Permissions.SampleCollect,      // B2 · BS tự lấy mẫu (PK ít người)
            Permissions.HardcopyReceive,    // B7 · nhận bản cứng
            Permissions.SidPrint, Permissions.ResultRead,
        }),
        new("nurse", "Điều dưỡng", new[]
        {
            Permissions.OrderRead, Permissions.ResultRead,
            Permissions.SampleCollect,      // B2 · điều dưỡng lấy mẫu thay BS
            Permissions.HardcopyReceive,
            Permissions.SidPrint,
        }),
        new("retail", "Khách lẻ", new[]
        {
            Permissions.CatalogRead,
            Permissions.OrderCreate, Permissions.OrderRead, Permissions.ResultRead,
        }),
        new("lab", "Trưởng phòng XN", new[]
        {
            Permissions.CatalogRead, Permissions.CatalogPriceEdit,
            Permissions.DealRead, Permissions.DealApprove,
            Permissions.OrderRead,
            Permissions.SampleCollect, Permissions.SampleGather, Permissions.SampleReceive, Permissions.SampleQc,
            Permissions.HardcopyDeliver,
            Permissions.SidPrint, Permissions.ResultRead, Permissions.ResultUpload, Permissions.ResultVerify,
        }),
        new("admin", "Quản trị hệ thống", Permissions.All),
    };
}

File: src/lablink-web/src/auth/permissions.ts
/** Permission hạt nhỏ dạng "<đối tượng>.<hành động>" (README · RBAC).
 * Menu và nút hành động render theo permission — KHÔNG hard-code theo role. */
export type Permission =
  | "catalog.read"
  | "catalog.price.edit"
  | "deal.create"
  | "deal.read"
  | "deal.approve"
  | "order.create"
  | "order.read"
  | "order.read.all"
  | "sample.collect"
  | "sample.gather"
  | "sample.receive"
  | "sample.qc"
  | "hardcopy.deliver"
  | "hardcopy.receive"
  | "sid.print"
  | "result.read"
  | "result.upload"
  | "user.manage"
  | "role.manage"
  | "department.manage"
  | "audit.read"
  | "report.read";

/** Nhóm (role) — user thuộc nhóm nào sẽ có tập permission tương ứng.
 * Bản thật: nạp từ server (GET /api/me). Đây là ánh xạ mẫu theo README. */
export type Role =
  | "doctor" | "nurse" | "retail" | "lab"
  | "fastlab_gather" | "fastlab_receive" | "fastlab_hardcopy"
  | "admin";

export const ROLE_LABEL: Record<Role, string> = {
  doctor: "Bác sĩ chỉ định",
  nurse: "Điều dưỡng",
  retail: "Khách lẻ",
  lab: "Trưởng phòng XN",
  fastlab_gather: "FastLab · Gom mẫu",
  fastlab_receive: "FastLab · KTV nhận mẫu",
  fastlab_hardcopy: "FastLab · Giao bản cứng",
  admin: "Quản trị hệ thống",
};

export const ROLE_PERMISSIONS: Record<Role, Permission[]> = {
  doctor: ["catalog.read", "deal.create", "deal.read", "order.create", "order.read", "sample.collect", "hardcopy.receive", "sid.print", "result.read"],
  nurse: ["order.read", "result.read", "sample.collect", "hardcopy.receive", "sid.print"],
  retail: ["catalog.read", "order.create", "order.read", "result.read"],
  lab: [
    "catalog.read", "catalog.price.edit",
    "deal.read", "deal.approve",
    "order.read", "order.read.all", "sample.collect", "sample.gather", "sample.receive", "sample.qc",
    "hardcopy.deliver",
    "sid.print", "result.read", "result.upload",
  ],
  fastlab_gather: ["order.read", "order.read.all", "sample.gather"],
  fastlab_receive: ["order.read", "order.read.all", "sample.receive", "sample.qc", "sid.print", "result.read", "result.upload"],
  fastlab_hardcopy: ["order.read", "order.read.all", "hardcopy.deliver"],
  admin: [
    "catalog.read", "catalog.price.edit",
    "deal.read", "deal.approve", "deal.create",
    "order.read", "order.read.all", "order.create",
    "sample.collect", "sample.gather", "sample.receive", "sample.qc",
    "hardcopy.deliver", "hardcopy.receive",
    "sid.print", "result.read", "result.upload",
    "user.manage", "role.manage", "department.manage", "audit.read", "report.read",
  ],
};
